### PR TITLE
SOLR-15861: ConcurrentUpdateSolrClient should work with ManagedExecutorService

### PR DESCRIPTION
**Pull request by sammyhk**
https://issues.apache.org/jira/browse/SOLR-15861

# Description

ConcurrentUpdateSolrClient (and ConcurrentUpdateHttp2SolrClient) should work with JSR-236 javax.enterprise.concurrent.ManagedExecutorService. Currently due to the code calling the java.util.concurrent.ExecutorService lifecycle methods (e.g. isShutdown(), isTerminated()) which produce exception java.lang.IllegalStateException: Lifecycle operation not supported.

# Solution

Wrap with try-catch block when calling those lifecycle methods.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

**Review by risdenk — CHANGES_REQUESTED** [review]
Probably makes sense to put this logic in https://github.com/apache/solr/blob/main/solr/solrj/src/java/org/apache/solr/common/util/ExecutorUtil.java?

**Review by risdenk — APPROVED** [review]
Merged main, added ExecutorUtil methods for isShutdown and isTerminated, updated usages of isShutdown and isTerminated to use ExecutorUtil.

**Review by dsmiley — COMMENTED** [review]
This is only really applicable to SolrJ where the client might provide an executor; right?  So I have some doubt about making a sweeping change across the codebase for a fringe case only in SolrJ.

Regardless, I feel we need someone to truly test this solution with the intended functionality to see it actually works as intended.

**Comment by risdenk** [comment]
> This is only really applicable to SolrJ where the client might provide an executor; right? So I have some doubt about making a sweeping change across the codebase for a fringe case only in SolrJ.

so maybe. In the "normal" case nothing changes here - `isShutdown()` is called just like it would have been. This simplifies everywhere we use an executor to check isShutdown or isTerminated to use `ExecutorUtil`. I can think of a few reasons this is beneficial:
* we can guarantee that all executors operate the same way
* we can add any logic to affect all the executors at once

Ideally when I went through this we should use `ExecutorUtil` `awaitTermination` and `shutdownAndAwaitTermination` in a few more places. This would handle `InterruptedException` and a few other things more nicely than we do today. Centralizing this in `ExecutorUtil` at least gives us a place to make these improvements instead of doing it all over the code base. Making these changes is going to be separate from this improvement.

> Regardless, I feel we need someone to truly test this solution with the intended functionality to see it actually works as intended.

Yea that would be nice. @sammyhk can you test this PR? have you tested these changes in your environment?

Even if we don't get this tested, we don't make things worse than they are today.

**Comment by risdenk** [comment]
@sammyhk if you get around to testing this that would be awesome. Please let us know how it goes. I'm going to merge this since it shouldn't make things worse.

File: solr/solrj/src/java/org/apache/solr/common/util/ExecutorUtil.java [hunk 1]
```diff
@@ -83,6 +83,24 @@ public interface InheritableThreadLocalProvider {
     void clean(AtomicReference<Object> ctx);
   }
 
+  public static boolean isShutdown(ExecutorService pool) {
+    try {
+      return pool.isShutdown();
+    } catch (IllegalStateException e) {
+      // JSR-239 ManagedExecutorService cannot query the lifecycle, so just return false
```
Comment (ryo0301) [inline solr/solrj/src/java/org/apache/solr/common/util/ExecutorUtil.java:90]: ❌ JSR-239
⭕️ JSR-236

Comment (risdenk) [reply solr/solrj/src/java/org/apache/solr/common/util/ExecutorUtil.java:90]: thanks fixing in a few minutes.

Comment (risdenk) [reply solr/solrj/src/java/org/apache/solr/common/util/ExecutorUtil.java:90]: Fixed in 8736c991bb6b80c11de06fbcb499ce8dbe2a7cb2

**Pull request merged**